scipy/optimize/_hessian_update_strategy.py

```diff
@@ -0,0 +1,307 @@
+               Second Edition (2006).
+        """
+        self.B = self._syr(1.0 / ys, y, a=self.B)
+        self.B = self._syr(-1.0 / sBs, Bs, a=self.B)
+
+    def update(self, delta_x, delta_grad):
+        if np.linalg.norm(delta_x) == 0.0:
+            return
+        if self.first_iteration:
+            self._scale_matrix(delta_x, delta_grad)
+            self.first_iteration = False
+        # Auxiliar variables w and z
+        if self.approx_type == 'hess':
+            w = delta_x
+            z = delta_grad
+        else:
+            w = delta_grad
+            z = delta_x
+        # Do some common operations
+        wz = np.dot(w, z)
+        Mw = self.dot(w)
+        wMw = Mw.dot(w)
+        if wMw == 0.0:
+            return
+        # Check if curvature condition is violated
+        if wz <= self.min_curvature * wMw:
+            # If the option 'skip_update' is set
+            # we just skip the update when the condion
+            # is violated.
+            if self.exception_strategy == 'skip_update':
+                return
+            # If the option 'damped_bfgs' is set we
+            # interpolate between the actual BFGS
+            # result and the unmodified matrix.
+            elif self.exception_strategy == 'damped_bfgs':
+                update_factor = (1-self.min_curvature) / (1 - wz/wMw)
+                z = update_factor*z + (1-update_factor)*Mw
+                wz = np.dot(w, z)
+                Mw = self.dot(w)
+                wMw = Mw.dot(w)
+        # Update matrix
+        if self.approx_type == 'hess':
+            self._update_hessian(wz, Mw, wMw, z)
+        else:
+            self._update_inverse_hessian(wz, Mw, wMw, z)
+
+
+class SR1(HessianUpdateStrategy):
+    """Symmetric-rank-1 Hessian update strategy.
+
+    Parameters
+    ----------
+    min_denominator : float
+        Define the minimum allowed value of the denominator
+        in the update. When the condition is violated we skip
+        the update. By default uses ``1e-8``.
+    init_scale : {float, 'auto'}, optional
+        Matrix scale at first iteration. At the first
+        iteration the Hessian matrix or its inverse will be initialized
+        with ``init_scale*np.eye(n)``, where ``n`` is the problem dimension.
+        Set it to 'auto' in order to use an automatic heuristic for choosing
+        the initial scale. The heuristic is described in [1]_, p.143.
+        By default uses 'auto'.
+
+    Notes
+    -----
+    The update is based on the description in [1]_, p.144-146.
+
+    References
+    ----------
+    .. [1] Nocedal, Jorge, and Stephen J. Wright. "Numerical optimization"
+           Second Edition (2006).
+    """
+
+    def __init__(self, min_denominator=1e-8, init_scale='auto'):
+        self.min_denominator = min_denominator
+        super(SR1, self).__init__(init_scale)
+
+    def update(self, delta_x, delta_grad):
+        if np.linalg.norm(delta_x) == 0.0:
```

Comment:
If these lines are assumed to be repeating throughout all derived implementations, maybe it's better to move it to base's `update` and derive `_update_implementation`? Like 

```
def update(...)
   if np.linalg.norm(delta_x) == 0.0:
      return
   if self.first_iteration:
       self._scale_matrix(delta_x, delta_grad)
       self.first_iteration = False
  self._update_implementation(...)
```

On the other hand, consider what I wrote about `self.init_scale`, maybe we should not really assume that there is some special scaling done on the first iteration.

Reply:
Ok, I created a method ``update`` that will be used by both ``BFGS`` and ``SR1``. The implementation is along the lines you have suggested.